feat: s3 로직 추가중

monda95 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: 
- 작업 요약: 피드백 받으며 s3 공용로직 추가 중

## 📄 상세 내용
- boto3 공식문서 참조하여 작성
- client 초기 생성 후 재사용방식, Presigned URL 생성, velidater, upload, delete 로직 추가
- moto_aws test 
- prefix 공용상수 참조하여 API 엔드포인트는  4개로 각 app의 view단에서 contect_type으로 경로 분기
 users groups, notes(images/files), recruitments,(images, files)

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

monda95 committed:
feat : s3_uploader 추가 (임시커밋)

# Conflicts:
#	apps/studies/urls.py

monda95 committed:
feat : s3_uploader 추가

- boto3 공식문서 참조하여 작성

- Presigned URL 생성, velidater, upload, delete 로직 추가

- s3_studies에 그룹, 노트 업로드 로직추가

# Conflicts:
#	apps/studies/urls.py

2joonkim commented on apps/core/utils/s3_uploader.py:
```diff
@@ -0,0 +1,199 @@
+        # 매핑 통합 조회
+        if ext in image_mime_by_ext:
+            if content_type not in image_mime_by_ext[ext]:
+                raise APIException("허용된 이미지 확장자/형식만 등록 가능합니다.")
```
현재 모든 에러를 APIException으로 던지고 있는거 같습니다!! 물론 의도적으로 명시를 해두셔서 문제는 없지만
클라이언트 잘못으로 에러가 나는 경우에는 ValidationError라던지 ParseError를 이용하여 던져주면
더 확실하게 식별이 가능할 거 같습니다!

monda95 committed:
add : boto3, boto3-stubs 추가

- 라이브러리에 각각 추가

# Conflicts:
#	poetry.lock
#	pyproject.toml

2joonkim commented on apps/core/utils/s3_uploader.py:
```diff
@@ -0,0 +1,199 @@
+###################################
+
+
+class S3Uploader:
```
저도 다몬님 코드보고 조금 찾아보고 확인해봤는데 이 방식은 클래스 import 시점에 실행이 되기 때문에
Django 설정이 완전히 로드되기 전에 import되면 settings 속성이 비어있을 수 있고 멀티프로세스 환경에서
실행이 안될 수 도 있는? 그런 문제점도 존재한다고 합니다

그래서 지연 초기화 형태로 바꿔주는게 안전하다고는 하는데 이 지연초기화라는 것이 양날의 검이기 때문에
현재는 이렇게 두고 코치님과 상의하셔서 진행해보시면 좋을 거 같습니다 ㅎㅎ

monda95 committed:
feat : test 추가

- moto_aws로 가상 s3 띄워서 테스트

- "file_url": f"{presigned_post['url']}{key}", 의 값이 가변할 수 있다고하여 고정값으로 변경 -> "file_url": cls.S3_BASE_URL + key,

monda95 committed:
fix : s3_client 생성(초기화) 방식 변경

- get_client 함수로thread-safe + Lazy Initialization 방식 채택

- 조교님 리뷰에서 django셋팅되기 전이나 설치 미완 관련 상황에서 미리 임포트되면서 발생할 수 있는문제들을 예방할 목적.

- 양날의 검이라고 언급하셨던 단점 중 "구니콘 + 유비콘" 사용중에 워커가 동시에 클라이언트를 만들어낼 수도 있다는 점들을, 예외처리와 쓰레드 세이프 방식으로 단일 생성 후 불변하게 유지

Meoyoug commented on apps/core/utils/s3_uploader.py:
```diff
@@ -0,0 +1,199 @@
+            "xlsx": {"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"},
+            "txt": {"text/plain"},
+            "csv": {"text/csv", "application/csv"},
+        }
```
요런거는 constants.py 나 base.py쪽에 상수값으로 빼두고 가져와서 사용하면 유지보수에 용이할것같습니다!

Meoyoug commented on apps/core/utils/s3_uploader.py:
```diff
@@ -0,0 +1,199 @@
+        raise APIException("허용된 확장자만 등록 가능합니다.")
+
+    @classmethod
+    def upload_file(cls, file: UploadedFile) -> str:
```
현재 이 메서드에서 이미지, 파일에 따른 S3 스토리지 내에서 경로 구분이없는 것 같아요

path를 파라미터로 넘겨받아서 key prefix로 연결해주면 좋을 것 같습니다.

예를 들어서 note의 이미지이면
```
uploads/note-images/
```

스터디 그룹의 이미지이면
```
uploads/group-images/
```

이런식으로요!

마지막에 키값은 그럼 f"{prefix}{uuid.uuid4().hex}.{ext}"가 되겠죠!

Meoyoug commented on apps/studies/views/s3_studies.py:
S3Uploader 코멘트에 따라 변경되면

1. 여기 각 View를 image, file 업로드에 대한 presigned url 발급 View로 각각 분리 
2. content_type에 따라서 이미지인지, 파일인지를 구분하고 S3Uploader.upload_file 메서드 호출 시에 key prefix로 전달 

둘중하나 선택해서 작업 진행하시면 될듯합니다.

monda95 committed:
fix : isorted

Meoyoug commented on apps/studies/views/s3_studies.py:
```diff
@@ -0,0 +1,151 @@
+                )
+
+        files_typed = cast(List[PresignFile], raw_files)  # mypy: List[PresignFile]로 확정
+        files = cast(List[Dict[str, str]], files_typed)  # S3Uploader 시그니처(list[dict[str,str]])와 호환
```
이건 시리얼라이저로 수행하면 좋을 것 같아요!

Meoyoug commented on apps/studies/views/s3_studies.py:
```diff
@@ -0,0 +1,151 @@
+                        },
+                    },
+                    status=status.HTTP_400_BAD_REQUEST,
+                )
```
여기도 마찬가지로 시리얼라이저로 수행

monda95 committed:
feat: prefix, MIME 상수 분리

monda95 committed:
fix: 이름변경 및 수정

- s3_presign.py 로 이름 변경

- group,note 모두 prefix 상수값 불러오고 note는 contect_type 기반으로 image/file 분기

- 5팀 Response 컨벤션 지키기위해 raise 대신 Response 사용 및 규칙유지

- urls.py presigndurl 엔드포인트 이름 변경

- view에서 하던 갯수 검증 책임 부분 시리얼라이저로 이동 + 외부입력하는 객체의 이름, 확장자업로드 전 단계에서 검증추가

- s3_uploader.py 상수값 불러오게끔 수정, 에러타입 수정, 프리픽스 + uuid 로 key 값 설정 등

- test_s3_uploader.py 수정된 코드에 맞춰 리팩터링

monda95 committed:
feat : add moto --dev

monda95 committed:
fix : s3_client로 롤백

- get_client 제거 및 롤백 , 단일/복수 삭제 , 제너레이트 url 부분도 맞춰서 롤백

- url 끝에 '/' 어펜드 슬래쉬 트루 설정을 안 할 예정인건지 혹은 프론트에서 통일을 요청한건지, url 끝에 '/'는 모두 지워달라는 조교님 요청이 있어, 미리 지우고 커밋

monda95 committed:
fix : s3 view 수정

-validate를 호출하는 형태로 불필요한 벨리데이트부분 제거

-validate_file_content_type 추가 views에서 ext로 하던 분기를 MIME 상수값으로 분기

- @extend_chema 활용하여 주석으로만 처리하던 부분을 문서화 양식에 더욱 알맞게 수정

- 상수값 저장파일 이름변경

monda95 committed:
fix : s3 view 수정

- s3_uploader에서 validate로 같이 검증하는 형태 수정

-validate_file_content_type 추가 views에서 ext로 하던 분기를 MIME 상수값으로 분기

- @extend_chema 활용하여 주석으로만 처리하던 부분을 문서화 양식에 더욱 알맞게 수정

- 상수값 저장파일 이름변경

- view note에서는 파일, 이미지 두가지 가능성이 있기 때문에 모든 벨리데이트 호출

monda95 committed:
fix : test 수정

- view단으로 검증 분리 이후 기준으로 벨리데이터 직접호출하며 예외발생확인
